#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QPixmap>

MainWindow::MainWindow(QWidget *parent) // –∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    socket = new QTcpSocket(this); // —Å–æ–∑–¥–∞–Ω–∏–µ —Å–æ–∫–µ—Ç–∞


    connect(socket, &QTcpSocket::readyRead, this, &MainWindow::slotReadyRead);
    connect(socket, &QTcpSocket::disconnected, socket, &QTcpSocket::deleteLater);

    ui->pushButton_2->setEnabled(false); // –∑–∞–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –∫–Ω–æ–ø–∫—É –æ—Ç–ø—Ä–∞–∫–∏
    ui->pushButton_3->setEnabled(false); // –∑–∞–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –∫–Ω–æ–ø–∫—É –æ—Ç–≤–µ—Ç–∞

    ui->statusbar->showMessage("Offline"); // –ø—Ä–∏ –∑–∞–ø—É—Å–∫–µ –≤—ã–≤–æ–¥ –≤ —Å—Ç–∞—Ç—É—Å –±–∞—Ä —Ç–µ–∫—Å—Ç –æ—Ñ—Ñ–ª–∞–π–Ω

    nextBlockSize = 0;
    QPushButton* smileButtons[] = { // –º–∞—Å—Å–∏–≤ —É–∫–∞–∑–∞—Ç–µ–ª–µ–π –∫–Ω–æ–ø–æ–∫ –≤—Å–µ—Ö —Å–º–∞–π–ª–∏–∫–æ–≤
        ui->smileButton_0, ui->smileButton_1, ui->smileButton_2,
        ui->smileButton_3, ui->smileButton_4, ui->smileButton_6,
        ui->smileButton_7, ui->smileButton_8, ui->smileButton_9,
        ui->smileButton_10, ui->smileButton_11, ui->smileButton_12,
        ui->smileButton_13, ui->smileButton_14, ui->smileButton_15
    };

    for (auto sml_btns : smileButtons) { // range-based —Ü–∏–∫–ª, –∫–æ—Ç–æ—Ä—ã–π –ø—Ä–æ—Ö–æ–¥–∏—Ç—å—Å—è –ø–æ –º–∞—Å—Å–∏–≤—É —Å–º–∞–π–ª–∏–∫–æ–≤ –∏ –ø—Ä–∏—Å–æ–π–¥–µ–Ω—è–µ—Ç –∫–∞–∂–¥—ã–π –∫ —Å–ª–æ—Ç—É sendEmoji()
        connect(sml_btns, SIGNAL(clicked()), this, SLOT(sendEmoji()));
    }
    connect(ui->smileButton_15, SIGNAL(clicked()), this, SLOT(sendEmoji()));

    QPixmap pix(":/img/img/hovrah_chat_logo.png"); // —ç—Ç–æ –¥–ª—è –∏–∫–æ–Ω–∫–∏
    int w = 150;
    int h = 150;

    ui->image->setPixmap(pix.scaled(w, h, Qt::KeepAspectRatio));

}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::on_pushButton_clicked() // –∫–Ω–æ–ø–∫–∞ connect
{
    if (!ui->name_line->text().isEmpty() && !ui->ip_line->text().isEmpty()){ // –µ—Å–ª–∏ –∏–º—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è –∏ ip –Ω–µ –ø—É—Å—Ç–æ–µ

        //socket->connectToHost("127.0.0.1", 2323); // –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–µ–≤–µ—Ä—É —á–µ—Ä–µ–∑ –ª–æ–∫–∞–ª—å–Ω—ã–π IP –∞–¥—Ä–µ—Å—Å –∏ —Ö–æ—Å—Ç

        socket->connectToHost(ui->ip_line->text(), 2323); //  –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–µ–≤–µ—Ä—É —á–µ—Ä–µ–∑ –°–û–ë–°–¢–í–ï–ù–ù–´–ô IP –∞–¥—Ä–µ—Å—Å –∏ —Ö–æ—Å—Ç

        qDebug() << "connect_button_pressed";
        SendToServer(ui->name_line->text(), "<span style='color:green;'> Joined the chat   üëã</span> ");
        ui->pushButton->setEnabled(false); // –∑–∞–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –ø–æ–≤—Ç–æ—Ä–Ω–æ–µ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ
        ui->name_line->setEnabled(false); // –∑–∞–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –ø–æ–ª–µ –∏–º–µ–Ω–∏
        ui->ip_line->setEnabled(false); // –∑–∞–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –ø–æ–ª–µ ip
        ui->pushButton_2->setEnabled(true); // —Ä–∞–∑–±–ª–æ–∫–∏—Ä–æ–≤–∞—Ç—å –∫–Ω–æ–ø–∫—É –æ—Ç–ø—Ä–∞–∫–∏
        ui->statusbar->showMessage("Online " + ui->ip_line->text()); // –æ–Ω–ª–∞–π–Ω

    }else{
        QMessageBox::warning(this, "Error", "Enter your name and server IP!"); // –æ–∫–Ω–æ —Å –æ—à–∏–±–∫–æ–π
    }
}

void MainWindow::closeEvent(QCloseEvent *event){ // –µ—Å–ª–∏ –æ–∫–Ω–æ –∑–∞–∫—Ä—ã–ª–æ—Å—å
    SendToServer(ui->name_line->text() , "<span style='color:green;'> Left the chat   ‚Ü™Ô∏èüö™</span> " );
}

void MainWindow::SendToServer(QString usr_name, QString str){ //–æ–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ –º–µ—Ç–æ–¥–∞
    Data.clear();  // –æ—á–∏—Å—Ç–∫–∞?
    QDataStream out(&Data, QIODevice::WriteOnly); // –¥–∞–Ω–Ω—ã–µ –Ω–∞ –≤—ã–≤–æ–¥(–≤ –∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä–µ: –º–∞—Å—Å–∏–≤ –±–∞–π—Ç–æ–≤ –∏ —Ä–µ–∂–∏–º —Ä–∞–±–æ—Ç—ã)
    out.setVersion(QDataStream::Qt_5_15);
    out << quint16(0) << QTime::currentTime() << usr_name << str; // –∑–∞–ø–∏—Å—å —á–µ—Ä–µ–∑ out —Å—Ç—Ä–æ–∫—É, –≤—ã–¥–µ–ª–µ–Ω—ã–µ –±–∞–π—Ç—ã –¥–ª—è –±–ª–æ–∫–∞(2 –±–∞–π—Ç–∞)?, –∏–º—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è –∏ —Ç–µ–∫—É—â–µ–µ –≤—Ä–µ–º—è –≤ –º–∞—Å—Å–∏–≤ Data
    out.device()->seek(0); // –ø–æ–¥—Å—á–µ—Ç —Ä–∞–∑–∏–µ—Ä–∞ –±–ª–æ–∫–∞?
    out << quint16(Data.size() - sizeof(quint16)); // —Ä–∞–∑–º–µ—Ä —Å–æ–æ–±—â–µ–Ω–∏—è, –∫–∞–∫ —Ä–∞–∑–Ω–æ—Å—Ç—å –º–µ–∂–¥—É –¥–ª–∏–Ω–Ω–æ–π –≤—Å–µ–≥–æ –±–ª–æ–∫–∞ –∏ –≤—ã–¥–µ–ª–µ–Ω—ã–º–∏ 2 –±–∞–π—Ç–∞–º–∏?
    socket->write(Data); // –∑–∞–ø–∏—Å—å Data –≤ —Å–æ–∫–µ—Ç
    ui->lineEdit->clear(); // –æ—á–∏—Å—Ç–∫–∞ –ø–æ–ª—è –¥–ª—è –≤–≤–æ–¥–∞ —Å–æ–æ–±—â–µ–Ω–∏—è
}


void MainWindow::slotReadyRead(){ // –æ–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ —Å–ª–æ—Ç–∞
    QDataStream in(socket); // –¥–ª—è –ø—Ä–∏–µ–º–∞ –ø–µ—Ä–µ–¥–∞—á–∏ –¥–∞–Ω–Ω—ã—Ö
    in.setVersion(QDataStream::Qt_5_15); // –≤–µ—Ä—Å–∏—è QDataStream

    if (in.status() == QDataStream::Ok){
        /*QString str;
        in >> str;
        ui->textBrowser->append(str); // –¥–æ–±–∞–≤–ª–µ–Ω–∏–µ —Å—Ç—Ä–æ–∫–∏ –≤ textBrowser*/

        for(;;){ // –±–µ—Å–∫–æ–Ω–µ—á–Ω—ã–π —Ü–∏–∫–ª
            if(nextBlockSize == 0){ // –µ—Å–ª–∏ —Ä–∞–∑–º–µ—Ä –±–æ–∫–∞ –Ω–µ–∏–∑–≤–µ—Å—Ç–µ–Ω
                if(socket->bytesAvailable() < 2){ // –µ—Å–ª–∏ –¥–ª—è —á—Ç–µ–Ω–∏—è –¥–æ—Å—Ç—É–ø–Ω–æ –Ω–µ –º–µ–Ω—å—à–µ 2 –±–∞–π—Ç
                    break; // –≤—ã—Ö–æ–¥ –∏–∑ —Ü–∏–∫–ª–∞
                }
                in >> nextBlockSize;
            }
            if(socket->bytesAvailable() < nextBlockSize){ // –¥–∞–Ω–Ω—ã–µ –ø—Ä–∏—à–ª–∏ –Ω–µ –ø–æ–ª–Ω–æ—Å—Ç—å—é
                break;
            }
            QString str;
            QTime time;
            QString usr_name;
            QString msg;
            in >> time >> usr_name >> str;
            nextBlockSize = 0; // —á—Ç–æ–±—ã –º–æ–∂–Ω–æ –±—ã–ª–æ –¥–∞–ª—å—à–µ –ø—Ä–∏–Ω–∏–º–∞—Ç—å –Ω–æ–≤—ã–µ —Å–æ–æ–±—â–µ–Ω–∏—è
            ui->textBrowser->append("[" + time.toString() + "] "+ "<b>" + usr_name + "</b>" + str); // –¥–æ–±–∞–≤–ª–µ–Ω–∏–µ —Å—Ç—Ä–æ–∫–∏ –∏ –≤—Ä–µ–º–µ–Ω–∏ –≤ textBrowser
            ui->msg_list->addItem("[" + time.toString() + "] "+ "<b>" + usr_name + "</b>" + str);
            QList<QListWidgetItem *> items = ui->listWidget->findItems(usr_name, Qt::MatchExactly); //  —Å–ø–∏—Å–∫–æ–º —É–∫–∞–∑–∞—Ç–µ–ª–µ–π –Ω–∞ –Ω–∞–π–¥–µ–Ω–Ω—ã–µ usr_nam–µ`—ã –≤ listWidget
            if (items.isEmpty()) { // –µ—Å–ª–∏ —Å–ø–∏—Å–æ–∫ –¥–ª—è –ø–æ–∏—Å–∫–∞ –∏–º–µ–µ—Ç –ø–æ–≤—Ç–æ—Ä. –∏–º—è –∏ –ø–∞—Ä–∞–º–µ—Ç—Ä usr_name –æ—Ç–ª–∏—á–∞–µ—Ç—å—Å—è –æ—Ç –∏–º–µ–Ω–∏ —Ç–µ–∫—É—â–µ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è

                ui->listWidget->addItem(new QListWidgetItem(QIcon(":/img/img/user.png"), usr_name)); // –¥–æ–±–∞–≤–∏—Ç—å –µ–≥–æ –≤ listWidget
            }
        }
    }else{
        ui->textBrowser->append("read error");
    }
}


void MainWindow::on_pushButton_2_clicked(){ // –∫–Ω–æ–ø–∫–∞ send
    if (!ui->lineEdit->text().isEmpty()){ // –µ—Å–ª–∏ –ø–æ–ª–µ –≤–≤–æ–¥–∞ –Ω–µ –ø—É—Å—Ç–æ–µ
        SendToServer(ui->name_line->text() ,  " " + ui->lasname_line->text() + "<span style='color:gray;'> >> </span> " + ui->lineEdit->text()); // –æ—Ç–ø—Ä–∞–≤–∫–∞ –∏–º–µ–Ω–∏ –∏ —Ç–µ–∫—Å—Ç–∞ –∏–∑ line edit
    }
}

void MainWindow::sendEmoji(){
    QPushButton *button = (QPushButton * )sender(); //—É–∫–∞–∑–∞—Ç–µ–ª—å –Ω–∞ –≤—Å–µ –∫–Ω–æ–ø–∫–∏ —Å–º–∞–π–ª–∏–∫–æ–≤ (sender –¥–ª—è –∫–Ω–æ–ø–æ–∫ —Å–º–∞–π–ª–∏–∫–æ–≤)
    //SendToServer(ui->name_line->text() ,  " " + ui->lasname_line->text() + "<span style='color:gray;'> >> </span> " + button->text()); // –æ—Ç–ø—Ä–∞–≤–∫–∞ –∏–º–µ–Ω–∏ –∏ —Å–º–∞–π–ª–∏–∫–æ–≤
    ui->lineEdit->insert(button->text());
}



void MainWindow::on_pushButton_3_clicked() // –∫–Ω–æ–ø–∫–∞ –æ—Ç–≤–µ—Ç –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é
{
    ui->lineEdit->setText("<span style='color:blue;'>" + ui->listWidget->currentItem()->text() + "</span>"); // –¥–æ–±–∞–≤–∏—Ç—å –∫ —Å—Ç—Ä–æ–∫–µ —Å–æ–æ–±—â–µ–Ω–∏—è –∏–º—è –≤—ã–±—Ä–∞–Ω–æ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
}




void MainWindow::on_listWidget_itemSelectionChanged()
{
     ui->pushButton_3->setEnabled(true);
}


